internal/logstreamer/logstreamer.go

```
@@ -0,0 +1,129 @@
+package logstreamer
+
+import (
+	"context"
+	"io"
+	"log"
+	"sync"
+	"time"
+)
+
+const defaultBufSize = 1024
+
+type LogStreamer struct {
+	reader  io.ReadCloser
+	broker  *Broker
+	buf     *SyncBuf
+	mu      sync.Mutex
+	streams []chan []byte
+}
+
+func NewLogStreamer(reader io.ReadCloser) *LogStreamer {
+	broker := NewBroker()
+	sub := broker.Subscribe()
+	buf := NewSyncBuf(sub)
+
+	ls := &LogStreamer{
+		reader:  reader,
+		broker:  broker,
+		buf:     buf,
+		streams: []chan []byte{},
+	}
+
+	go ls.readLogs()
+	go broker.Broadcast()
+
+	return ls
+}
+
+func (s *LogStreamer) readLogs() {
+	for {
+		pack := make([]byte, defaultBufSize)
+		_, err := s.reader.Read(pack)
+		if err == io.EOF {
+			// There is nothing left to read now, but the task is still
+			// running - so just chill for a second and try again
+			time.Sleep(time.Second)
+			continue
+		} else if err != nil {
+			// We end up here in case the task is finished and the reader
+			// is closed. It might be that something unexpected happened,
+			// but anyway shutting down this goroutine as we aren't going
+			// anywhere from here
+			break
```

Can you return this unexpected error instead of just breaking?

---

Reworked error handling a bit. Added logging and writing the error to the buffer, so the error will be returned to the user if he asks for the logs